File: server/services/tradingOrchestrator.ts


import { freqtradeAdapter, jesseAdapter } from './tradingFrameworks';

type Prediction = { action: string; confidence: number; source?: string };

export class TradingOrchestrator {
  private started = false;

  public startAll(): void {
    if (this.started) return;
    try {
      freqtradeAdapter.start();
      jesseAdapter.start();
      this.started = true;
    } catch (e) {
      // ignore start failures, adapters may not be available
      this.started = false;
    }
  }

  public stopAll(): void {
    try {
      freqtradeAdapter.stop();
      jesseAdapter.stop();
    } finally {
      this.started = false;
    }
  }

  public async getEnsemblePrediction(payload: any): Promise<{ action: string; confidence: number; votes: Prediction[] }> {
    const votes: Prediction[] = [];

    // Always include a local 'none' baseline to avoid division by zero
    try {
      const f = await freqtradeAdapter.predict(payload).catch(() => null);
      if (f && f.action) votes.push({ action: f.action, confidence: typeof f.confidence === 'number' ? f.confidence : 0.5, source: 'freqtrade' });
    } catch (e) {
      // ignore
    }

    try {
      const j = await jesseAdapter.predict(payload).catch(() => null);
      if (j && j.action) votes.push({ action: j.action, confidence: typeof j.confidence === 'number' ? j.confidence : 0.5, source: 'jesse' });
    } catch (e) {
      // ignore
    }

    // If no external votes, return neutral
    if (votes.length === 0) {
      return { action: 'hold', confidence: 0, votes };
    }

    // Tally weighted votes
    const tally: Record<string, number> = {};
    let total = 0;
    for (const v of votes) {
      tally[v.action] = (tally[v.action] || 0) + v.confidence;
      total += v.confidence;
    }

    let bestAction = 'hold';
    let bestWeight = 0;
    for (const k of Object.keys(tally)) {
      if (tally[k] > bestWeight) {
        bestWeight = tally[k];
        bestAction = k;
      }
    }

    const normalizedConfidence = total > 0 ? bestWeight / total : 0;

    return { action: bestAction, confidence: normalizedConfidence, votes };
  }

  public async pingAll(): Promise<{ freqtrade?: any; jesse?: any }> {
    const out: { freqtrade?: any; jesse?: any } = {};
    try {
      out.freqtrade = await freqtradeAdapter.ping().catch((e) => ({ ok: false, error: String(e) }));
    } catch (e) {
      out.freqtrade = { ok: false, error: String(e) };
    }

    try {
      out.jesse = await jesseAdapter.ping().catch((e) => ({ ok: false, error: String(e) }));
    } catch (e) {
      out.jesse = { ok: false, error: String(e) };
    }

    return out;
  }

  public async backtest(payload: any): Promise<{ results: any[]; summary: { avgProfitPct: number; totalTrades: number } }> {
    const results: any[] = [];
    try {
      const f = await freqtradeAdapter.backtest(payload).catch(() => null);
      if (f) results.push({ source: 'freqtrade', ...f });
    } catch (e) {
      // ignore
    }

    try {
      const j = await jesseAdapter.backtest(payload).catch(() => null);
      if (j) results.push({ source: 'jesse', ...j });
    } catch (e) {
      // ignore
    }

    // compute simple summary
    let totalTrades = 0;
    let sumProfit = 0;
    for (const r of results) {
      if (typeof r.trades === 'number') totalTrades += r.trades;
      if (typeof r.profit_pct === 'number') sumProfit += r.profit_pct;
    }
    const avgProfitPct = results.length > 0 ? sumProfit / results.length : 0;

    return { results, summary: { avgProfitPct, totalTrades } };
  }
}

export const tradingOrchestrator = new TradingOrchestrator();

File: server_fastapi/services/trading_orchestrator.py
from typing import Dict, Any, List, Optional
from pydantic import BaseModel
import logging
import asyncio

logger = logging.getLogger(__name__)

class Prediction(BaseModel):
    action: str
    confidence: float
    source: Optional[str] = None

class EnsemblePrediction(BaseModel):
    action: str
    confidence: float
    votes: List[Prediction]

class PingResult(BaseModel):
    freqtrade: Optional[Dict[str, Any]] = None
    jesse: Optional[Dict[str, Any]] = None

class BacktestSummary(BaseModel):
    avg_profit_pct: float
    total_trades: int

class BacktestResult(BaseModel):
    results: List[Dict[str, Any]]
    summary: BacktestSummary

class TradingOrchestrator:
    def __init__(self):
        self.started = False
        # Mock adapters - in real implementation, import actual adapters
        self.freqtrade_adapter = None
        self.jesse_adapter = None

    def start_all(self) -> None:
        """Start all trading adapters"""
        if self.started:
            return

        try:
            # In real implementation:
            # self.freqtrade_adapter.start()
            # self.jesse_adapter.start()
            logger.info("Starting all trading adapters")
            self.started = True
        except Exception as e:
            logger.warning(f"Failed to start some adapters: {e}")
            self.started = False

    def stop_all(self) -> None:
        """Stop all trading adapters"""
        try:
            # In real implementation:
            # self.freqtrade_adapter.stop()
            # self.jesse_adapter.stop()
            logger.info("Stopping all trading adapters")
        finally:
            self.started = False

    async def get_ensemble_prediction(self, payload: Dict[str, Any]) -> EnsemblePrediction:
        """Get ensemble prediction from all available adapters"""
        votes: List[Prediction] = []

        # Mock predictions - in real implementation, call actual adapters
        try:
            # Simulate freqtrade prediction
            freqtrade_result = await self._mock_freqtrade_predict(payload)
            if freqtrade_result and freqtrade_result.get('action'):
                votes.append(Prediction(
                    action=freqtrade_result['action'],
                    confidence=freqtrade_result.get('confidence', 0.5),
                    source='freqtrade'
                ))
        except Exception as e:
            logger.warning(f"Freqtrade prediction failed: {e}")

        try:
            # Simulate jesse prediction
            jesse_result = await self._mock_jesse_predict(payload)
            if jesse_result and jesse_result.get('action'):
                votes.append(Prediction(
                    action=jesse_result['action'],
                    confidence=jesse_result.get('confidence', 0.5),
                    source='jesse'
                ))
        except Exception as e:
            logger.warning(f"Jesse prediction failed: {e}")

        # If no external votes, return neutral
        if not votes:
            return EnsemblePrediction(action='hold', confidence=0.0, votes=votes)

        # Tally weighted votes
        tally: Dict[str, float] = {}
        total = 0.0

        for vote in votes:
            action = vote.action
            confidence = vote.confidence
            tally[action] = tally.get(action, 0.0) + confidence
            total += confidence

        # Find best action
        best_action = 'hold'
        best_weight = 0.0

        for action, weight in tally.items():
            if weight > best_weight:
                best_weight = weight
                best_action = action

        normalized_confidence = best_weight / total if total > 0 else 0.0

        return EnsemblePrediction(
            action=best_action,
            confidence=normalized_confidence,
            votes=votes
        )

    async def _mock_freqtrade_predict(self, payload: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Mock freqtrade prediction - replace with actual adapter call"""
        await asyncio.sleep(0.1)  # Simulate async call
        # Mock logic based on payload
        return {'action': 'buy', 'confidence': 0.65}

    async def _mock_jesse_predict(self, payload: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Mock jesse prediction - replace with actual adapter call"""
        await asyncio.sleep(0.1)  # Simulate async call
        # Mock logic based on payload
        return {'action': 'hold', 'confidence': 0.55}

    async def ping_all(self) -> PingResult:
        """Ping all trading adapters"""
        result = PingResult()

        try:
            # In real implementation: result.freqtrade = await self.freqtrade_adapter.ping()
            result.freqtrade = {'ok': True, 'version': '1.0.0'}
        except Exception as e:
            result.freqtrade = {'ok': False, 'error': str(e)}

        try:
            # In real implementation: result.jesse = await self.jesse_adapter.ping()
            result.jesse = {'ok': True, 'version': '1.0.0'}
        except Exception as e:
            result.jesse = {'ok': False, 'error': str(e)}

        return result

    async def backtest(self, payload: Dict[str, Any]) -> BacktestResult:
        """Run backtest across all adapters"""
        results: List[Dict[str, Any]] = []

        try:
            # Mock freqtrade backtest
            freqtrade_result = await self._mock_freqtrade_backtest(payload)
            if freqtrade_result:
                results.append({'source': 'freqtrade', **freqtrade_result})
        except Exception as e:
            logger.warning(f"Freqtrade backtest failed: {e}")

        try:
            # Mock jesse backtest
            jesse_result = await self._mock_jesse_backtest(payload)
            if jesse_result:
                results.append({'source': 'jesse', **jesse_result})
        except Exception as e:
            logger.warning(f"Jesse backtest failed: {e}")

        # Compute summary
        total_trades = 0
        sum_profit = 0.0

        for result in results:
            if 'trades' in result and isinstance(result['trades'], (int, float)):
                total_trades += int(result['trades'])
            if 'profit_pct' in result and isinstance(result['profit_pct'], (int, float)):
                sum_profit += float(result['profit_pct'])

        avg_profit_pct = sum_profit / len(results) if results else 0.0

        return BacktestResult(
            results=results,
            summary=BacktestSummary(
                avg_profit_pct=avg_profit_pct,
                total_trades=total_trades
            )
        )

    async def _mock_freqtrade_backtest(self, payload: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Mock freqtrade backtest - replace with actual adapter call"""
        await asyncio.sleep(0.5)  # Simulate longer running operation
        return {
            'trades': 150,
            'profit_pct': 12.5,
            'win_rate': 0.55,
            'max_drawdown': 8.2
        }

    async def _mock_jesse_backtest(self, payload: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Mock jesse backtest - replace with actual adapter call"""
        await asyncio.sleep(0.5)  # Simulate longer running operation
        return {
            'trades': 120,
            'profit_pct': 15.3,
            'win_rate': 0.58,
            'max_drawdown': 6.8
        }

    async def get_user_bots(self, user_id: int) -> List[Dict[str, Any]]:
        """Get bots for a specific user"""
        # Mock implementation - in real implementation, query database
        return [
            {
                'id': 1,
                'user_id': user_id,
                'name': 'BTC Scalper',
                'status': 'running',
                'strategy': 'scalping',
                'symbol': 'BTC/USD',
                'last_update': asyncio.get_event_loop().time()
            },
            {
                'id': 2,
                'user_id': user_id,
                'name': 'ETH Holder',
                'status': 'stopped',
                'strategy': 'hold',
                'symbol': 'ETH/USD',
                'last_update': asyncio.get_event_loop().time()
            }
        ]

    async def get_bot_status(self, user_id: int, bot_id: int) -> Dict[str, Any]:
        """Get status of a specific bot"""
        bots = await self.get_user_bots(user_id)
        for bot in bots:
            if bot['id'] == bot_id:
                return bot
        return {'error': 'Bot not found'}

# Global instance
trading_orchestrator = TradingOrchestrator()
